CRM/Volunteer/BAO/Project.php

```diff
@@ -237,59 +237,166 @@ static function isActive($entityId, $entityTable) {
+    }
+    $dao->free();
+
+    return $result;
+  }
+
+  /**
+   * Helper method to filter Projects by related contact.
+   *
+   * Conditionally invoked by CRM_Volunteer_BAO_Project::retrieve().
+   *
+   * @param array $projectContacts
+   *   @see CRM_Volunteer_BAO_Project::create() for details on this parameter
+   * @return mixed
+   *   Boolean FALSE if no projects have the specified contact relationships;
+   *   String SQL fragment otherwise
+   */
+  private static function buildContactJoin(array $projectContacts) {
+    $result = FALSE;
+    $onClauses = array();
+
+    $relTypes = CRM_Core_OptionGroup::values(
+      CRM_Volunteer_BAO_ProjectContact::RELATIONSHIP_OPTION_GROUP,
+      TRUE, FALSE, FALSE, NULL, 'name');
 
-    while ($project->fetch()) {
-      $result[(int) $project->id] = clone $project;
+    foreach ($projectContacts as $relType => $contactIds) {
+      if (!CRM_Utils_Type::validate($relType, 'Integer', FALSE)) {
+        $relType = $relTypes[$relType];
+      }
+      $contactIds = implode(',', (array) $contactIds);
+
+      $onClauses[] = "(vpc.contact_id IN ($contactIds) AND vpc.relationship_type_id = $relType)";
     }
 
-    $project->free();
+    if (count($onClauses)) {
+      $strOnClauses = implode(' OR ', $onClauses);
+      $result = "INNER JOIN `civicrm_volunteer_project_contact` vpc
+        ON vp.id = vpc.project_id AND ($strOnClauses)";
+    }
 
     return $result;
   }
 
   /**
+   * Helper method to filter Projects by location.
+   *
+   * @param array $params
+   *   <ol>
+   *     <li>string city - optional. Not used in this function, just passed along for geocoding.</li>
+   *     <li>mixed country - required if lat/lon not provided. Can be country_id or string.</li>
+   *     <li>float lat - required if country not provided</li>
+   *     <li>float lon - required if country not provided</li>
+   *     <li>string postal_code - optional. Not used in this function, just passed along for geocoding.</li>
+   *     <li>float radius - required</li>
+   *     <li>string street_address - optional. Not used in this function, just passed along for geocoding.</li>
+   *     <li>string unit - optional, defaults to meters unless 'mile' is specified</li>
+   *   </ol>
+   * @return string
+   *   SQL fragment (partial where clause)
+   * @throws Exception
+   */
+  private static function buildProximityWhere(array $params) {
+    $country = $lat = $lon = $radius = $unit = NULL;
+    extract($params, EXTR_IF_EXISTS);
+
+    // ensure that radius is a float
+    if (!CRM_Utils_Rule::numeric($radius)) {
+      throw new Exception(ts('Radius should exist and be numeric'));
+    }
+
+    if (!CRM_Utils_Rule::numeric($lat) || !CRM_Utils_Rule::numeric($lon)) {
+      if (empty($country)) {
+        throw new Exception(ts('Either Country or both Latitude and Longitude are required'));
+      }
+
+      // TODO: I think CRM_Utils_Geocode_*::format should be responsible for this
```

Comment:
Per your request, I created this: [CRM-17245](https://issues.civicrm.org/jira/browse/CRM-17245).